### DISCUSSION TITLE
Issue while saving the audit data into another database particularly during a Transaction scenario

### DISCUSSION THREAD

**Discussion by ragashreeinfineon** [thepirat000/Audit.NET#712] [Q&A]
I have an Application Db and audit Db, i want to log the audit records of Application into this audit Db. But my configuration fails during a transaction scenario saying multiple connections can't access same transaction

```csharp
public class AppDbContext : AuditDbContext
{
     public AppDbContext(DbContextOptions<AppDbContext> options) : base(options)
     {

      }
      // my dbsets here
 }
 public class AuditingDbContext : DbContext
{
     public AuditingDbContext(DbContextOptions<AuditingDbContext> options) : base(options)
     {

      }
      // i have only one AuditLogs table here
 }
```

 In program.cs :

```csharp
 Audit.Core.Configuration.Setup()
    .UseEntityFramework(ef => ef
    .UseDbContext<AuditingDbContext>(_options)
    .AuditTypeMapper(t => typeof(AuditLog))
    .AuditEntityAction<AuditLog>((ev, entry, entity) =>
    {
        AuditConfiguration.AddAudit(entry, entity);   // sepatate class where it just assigns the required value
    })
    .IgnoreMatchedProperties(true));
    
    // the above works in saving records into my audit database, if Transaction is not involved.
```
    
So changes made to make it work:
    When AuditDataProvider is set in program class it did not work as expected, instead setting audit provider in the AppDbContext stores the whole record into the Db as a single row but I want them as individual rows.. tried many ways but still no luck!

**Comment by thepirat000** [comment]
Could you share the code used to create the transaction or where the error occurs? Additionally, please provide the exception details, including the message and call stack trace

**Reply by ragashreeinfineon** [reply]
Since i can't directly mention some classes and method names i have modified them into general names

audit config made in program .cs file:

```c#
Audit.Core.Configuration.Setup().UseMySql(config =>
{
    config.ConnectionString(GetDatabaseConnectionString(builder, "auditDb"));
});

var _options = new DbContextOptionsBuilder<AppAuditContext>()
    .UseMySQL(GetDatabaseConnectionString(builder, "auditDb")).Options;
Audit.Core.Configuration.Setup()
.UseEntityFramework(ef => ef
.UseDbContext< AppAuditContext >(_options)
.AuditTypeMapper(t => typeof(AuditLog))
.AuditEntityAction<AuditLog>((ev, entry, entity) =>
{
    AuditConfiguration.AddAudit(entry, entity);
})
.IgnoreMatchedProperties(true));
```

Config Done with AuditConfiguration.AddAudit(entry, entity) method ->

```c#
public static void AddAudit(EventEntry entry, AuditLog entity)
{
    entity.Id = Guid.NewGuid();
    entity.Entity = entry.Table;
    entity.EntityId = entry.PrimaryKey.Select(x => new Guid(x.Value.ToString())).ToList();
    entity.Action = entry.Action;
    entity.ChangedAt = DateTime.UtcNow;
    if (string.Equals(entity.Action, "Insert"))
    {
        entity.NewValue = entry.ColumnValues.Select(x => $"{x.Key} : {x.Value}");
        entity.ChangedBy = entry.ColumnValues.Where(kvp => kvp.Key == "CreatedBy").Select(kvp => kvp.Value)?.FirstOrDefault()?.ToString();
    }
    else if (string.Equals(entity.Action, "Update"))
    {
        string? isActive = entry.ColumnValues.Where(kvp => kvp.Key == "IsActive").Select(kvp => kvp.Value)?.FirstOrDefault()?.ToString();
        if (!string.IsNullOrEmpty(isActive) &&
            bool.TryParse(isActive, out bool result) && result)
        {
            entity.NewValue = entry.Changes.Select(x => $"{x.ColumnName} : {x.NewValue}");
            entity.OldValue = entry.Changes.Select(x => $"{x.ColumnName} : {x.OriginalValue}");
            entity.ChangedBy = entry.ColumnValues.Where(kvp => kvp.Key == "UpdatedBy").Select(kvp => kvp.Value).FirstOrDefault().ToString();
        }
        else
        {
            entity.Action = "Delete";
            entity.OldValue = entry.Changes.Select(x => $"{x.ColumnName} : {x.OriginalValue}");
            entity.ChangedBy = entry.ColumnValues.Where(kvp => kvp.Key == "DeletedBy").Select(kvp => kvp.Value).FirstOrDefault().ToString();
        }
    }
    // only for delete operation -
    else
    {
        entity.OldValue = entity.OldValue = entry.Changes.Select(x => $"{x.ColumnName} : {x.OriginalValue}");
        entity.ChangedBy = entry.ColumnValues.Where(kvp => kvp.Key == "DeletedBy").Select(kvp => kvp.Value).FirstOrDefault().ToString();
    }
}
```

Config Done in AppContext.cs file :

```c#
public class AppContext: AuditDbContext, IAppContext{
	// all the dbsets here
	public AppContext (DbContextOptions<AppContext> options)
     : base(options)
{
    // injected the auditdataprovider here like a constructor injection by first configuring it in program file and adding that here.. see below
}
}
```

Config Done in AppAuditContext.cs file :

```c#
public class AppAuditContext : DbContext,IAppAuditContext
{
    public AppAuditContext (DbContextOptions< AppAuditContext > options) : base(options)
    {

    }
    public DbSet<AuditLog> AuditLogs { get; set; } // only single table in this db and it has a separate database created for audit purpose only
    // onmodel configuration here
}
```

Code where we get error :

```c#
public async Task<xdto > Create(xdto newRecord, string createdBy)
{
    await DeleteyrecordBy(newRecord.EId); // this method has savechangesasync once the record is deletd

    // some logic

    var newRequests = CreateRequest(newRecord, createdBy);
    _appContext.[app_tablename].AddRange(newRequests);

    // some logic here

    if (something is true)
    {
        // getentity logic based on an Id
        // internal updateMethod(with some params) // this method updated the entity fetched from the above getentity and this doeasn’t have any savechangesasync inside it and done inside the transactionscope
        // some logic here
    }
    else
    {
        
        // same update method implementation for different entity
    }

    try
    {
        using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled); // Enable async flow
        await _appContext.SaveChangesAsync().ConfigureAwait(true); // before configuring auditdataprovider this line calls the AddAudit method configured in the program class, suppose in this transaction if 2 entries are there at the end of processing the second entry it suddenly displays “multiple connections trying to use same transactions..” for now I could not give full error bcoz on configuring auditdataprovider in the program class and injecting that into our appContext class constructor the error is gone but the other issue is, it saves both the entries in the same row where sometimes we might want to filter the records based on entities.. in that case we won’t get proper record
        scope.Complete();
        // some logic here
    }
    catch (TransactionAbortedException ex)
    {
        throw new DbUpdateException(newRecord.EntityId.ToString(), ex);
    }
    return something;
}
```

Auditdtaaprovider config to fix the error and to provide work around:
In Program class 

```c#
//Audit.Core.Configuration.Setup().UseMySql(config =>
//{
//    config.ConnectionString(GetDatabaseConnectionString(builder, "auditDb"));
//});

builder.Services.AddSingleton<AuditDataProvider>(new MySqlDataProvider(config => config
.ConnectionString(GetDatabaseConnectionString(builder, "auditDb"))
.TableName("AuditLogs")
.IdColumnName("Id")
.JsonColumnName("Data")
.CustomColumn("Id", ev => Guid.NewGuid().ToString())
.CustomColumn("EntityId", entry => string.Join(",",entry.GetEntityFrameworkEvent().Entries.SelectMany(x => x.PrimaryKey.Select(kv => new Guid(kv.Value.ToString()))).ToList()))
.CustomColumn("Entity", entry => string.Join(",", entry.GetEntityFrameworkEvent().Entries.Select(x => x.Table.ToString())))
.CustomColumn("Action", entry => string.Join(",",entry.GetEntityFrameworkEvent().Entries.Select(x => x.Action.ToString())))
.CustomColumn("ChangedAt", entry => DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss"))
.CustomColumn("ChangedBy", entry => entry.Environment.UserName)
));
// these custom columns are bit difficult to manipulate according to the table schema we want

// When using audit data provider this add audit is not triggered.. but before adding audit data provider this Addaudit was executed..
var _options = new DbContextOptionsBuilder<AppAuditContext>()
    .UseMySQL(GetDatabaseConnectionString(builder, "auditDb")).Options;
Audit.Core.Configuration.Setup()
.UseEntityFramework(ef => ef
.UseDbContext<AppAuditContext>(_options)
.AuditTypeMapper(t => typeof(AuditLog))
.AuditEntityAction<AuditLog>((ev, entry, entity) =>
{
    AuditConfiguration.AddAudit(entry, entity);
})
.IgnoreMatchedProperties(true));
```

In appcontext we inject this in constructor but it inserts the transaction inside a single row but not as individual row entries for each entity

**Comment by thepirat000** [comment]
It's still unclear to me what you're trying to achieve. I notice multiple calls to `Audit.Core.Configuration.Setup().UseXXXX()`, which doesn't make sense—only one Audit Data Provider can be configured at a time. Are you attempting to use the `MySqlDataProvider` or the `EntityFrameworkDataProvider`?  

Are both databases hosted on the same server?

I tried reproducing the issue with the following code, and it works as expected, but I'm running both databases (app and audit) on the same MS SQL Server, just in different databases:

```c#
TransactionManager.ImplicitDistributedTransactions = true;

using var transactionScope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);

ctx.Values.Add(entity);

await ctx.SaveChangesAsync();

transactionScope.Complete();
```

```c#
builder.Services.AddDbContext<MyContext>(_ => _.UseSqlServer("data source=localhost;initial catalog=TestTran1;integrated security=true;Encrypt=False"));
builder.Services.AddDbContext<MyAuditContext>(_ => _.UseSqlServer("data source=localhost;initial catalog=TestTran2;integrated security=true;Encrypt=False"));

Audit.Core.Configuration.Setup()
   .UseEntityFramework(d => d
       .UseDbContext(ev => ev.GetEntityFrameworkEvent().GetDbContext().GetService<MyAuditContext>())
       .AuditTypeMapper(t => typeof(AuditLog))
       .AuditEntityAction<AuditLog>((ev, entry, auditLog) =>
       {
           auditLog.Table = entry.Table;
           auditLog.ChangeType = entry.Action;
           auditLog.Value = entry.ToJson();
       })
       .IgnoreMatchedProperties(true));
```

**Reply by ragashreeinfineon** [reply]
{
  "type": "https://tools.ietf.org/html/rfc9110#section-15.6.1",
  "title": "System.NotSupportedException",
  "status": 500,
  "detail": "Multiple simultaneous connections or connections with different connection strings inside the same transaction are not currently supported.",
  "exception": {
    "details": "System.NotSupportedException: Multiple simultaneous connections or connections with different connection strings inside the same transaction are not currently supported.\r\n   at MySql.Data.MySqlClient.Interceptors.ExceptionInterceptor.Throw(Exception exception)\r\n   at MySql.Data.MySqlClient.MySqlConnection.Throw(Exception ex)\r\n   at MySql.Data.MySqlClient.MySqlConnection.OpenAsync(Boolean execAsync, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternalAsync(Boolean errorsExpected, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternalAsync(Boolean errorsExpected, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenAsync(CancellationToken cancellationToken, Boolean errorsExpected)\r\n   at MySql.EntityFrameworkCore.Storage.Internal.MySQLRelationalConnection.OpenAsync(CancellationToken cancellationToken, Boolean errorsExpected)\r\n   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)\r\n   at MySql.EntityFrameworkCore.Storage.Internal.MySQLExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)\r\n   at Audit.EntityFramework.Providers.EntityFrameworkDataProvider.InsertEventAsync(AuditEvent auditEvent, CancellationToken cancellationToken)\r\n   at Audit.EntityFramework.Providers.EntityFrameworkDataProvider.InsertEventAsync(AuditEvent auditEvent, CancellationToken cancellationToken)\r\n   at Audit.Core.AuditScope.SaveEventAsync(Boolean forceInsert, CancellationToken cancellationToken)\r\n   at Audit.Core.AuditScope.SaveAsync(CancellationToken cancellationToken)\r\n   at Audit.EntityFramework.DbContextHelper.SaveScopeAsync(IAuditDbContext context, IAuditScope scope, EntityFrameworkEvent event, CancellationToken cancellationToken)\r\n   at Audit.EntityFramework.DbContextHelper.SaveChangesGetAuditAsyncImpl(IAuditDbContext context, Func`1 baseSaveChanges, CancellationToken cancellationToken)\r\n   at Audit.EntityFramework.DbContextHelper.SaveChangesAsync(IAuditDbContext context, Func`1 baseSaveChanges, CancellationToken cancellationToken)\r\n   at Audit.EntityFramework.AuditDbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)\r\n   at xxx
  
  }
}

this the error I'm facing - Actually i'm trying to save the audits of my entities in different database dedicated for audit purpose. we have 2 different Db connection string : one is for appdb and another is for audit db.. so we created two dbcontexts class in our application
The appDbcontext is inherited from the AuditDbContext and another AppAuditContext is inherited from DbContext.. 
First my configurations were simple like the one you provided but it did not work, could you please let me know if you are using this AppAuditDbContext class inside the AppDbContext
I first tried to use efdataprovider then switched to mysqldataprovider. So with "mysqldataprovider" as my auditdataprovider i did not receive any error but if the transaction is updating two or more tables then i want to know how can i insert them into my db separately as individual rows.

**Comment by thepirat000** [comment]
This seems to be a limitation for MySQL. Can you try with the suggestion in this [SO answer](https://stackoverflow.com/a/35985897/122195)?

**Reply by ragashreeinfineon** [reply]
I tried making it false but git the below error
"title": "System.InvalidOperationException",
  "status": 500,
  "detail": "A root ambient transaction was completed before the nested transaction. The nested transactions should be completed first.",
  "exception": {
    "details": "System.InvalidOperationException: A root ambient transaction was completed before the nested transaction. The nested transactions should be completed first.\r\n   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.HandleTransactionCompleted(Object sender, TransactionEventArgs e)\r\n   at System.Transactions.TransactionStateAborted.EnterState(InternalTransaction tx)\r\n   at System.Transactions.Transaction.Rollback()\r\n   at System.Transactions.TransactionScope.InternalDispose()\r\n   at System.Transactions.TransactionScope.Dispose()
    
    For now i was able to save record my using mysqldataprovider as the auditdataprovider, below is the config in program file
    builder.Services.AddSingleton<AuditDataProvider>(new MySqlDataProvider(config => config
.ConnectionString(GetDatabaseConnectionString(builder, "auditDb"))
.TableName("AuditLogs")
.IdColumnName("Id")
.JsonColumnName("Data")
.CustomColumn("Id", ev => Guid.NewGuid().ToString())
.CustomColumn("EntId", entry => string.Join(",", entry.GetEntityFrameworkEvent().Entries.SelectMany(x => x.PrimaryKey.Select(kv => new Guid(kv.Value.ToString()))).ToList()))
.CustomColumn("Entity", entry => string.Join(",", entry.GetEntityFrameworkEvent().Entries.Select(x => x.Table.ToString())))
.CustomColumn("Action", entry => string.Join(",", entry.GetEntityFrameworkEvent().Entries.Select(x => x.Action.ToString())))
.CustomColumn("ChangedAt", entry => DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss"))
.CustomColumn("ChangedBy", entry => entry.Environment.UserName)
));
if some transaction has updates to two tables for example tableA and tableB then it saves them in one single row like this -> row1 : [tableA,tableB] but it has to be stored like row1 : tableA, and for another record it should create row2 and store them individually.
Roght now i dont see a way to loop through and save them separately. bcoz whenever savechangesasync is hit, it just saves them whole bunch of record

**Comment by thepirat000** [comment]
If you use the MySqlDataProvider, the audit logs will not be inserted within the same transaction as the application database. In this case, I recommend using the EFDataProvider but avoid creating a TransactionScope.

This limitation lies within the MySQL server and is unrelated to this library. My understanding is that TransactionScope will function across different connections only if your MySQL server supports XA transactions, which are available exclusively with the InnoDB engine.
